src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Instance/InstanceDiscovery.cs

```diff
@@ -0,0 +1,153 @@
+            "login-us.microsoftonline.com", // Microsoft Azure US Government - Legacy
+            "login.microsoftonline.us", // Microsoft Azure US Government
+            "login.microsoftonline.com" // Microsoft Azure Worldwide
+        });
+
+        public const string AuthorizeEndpointTemplate = "https://{host}/{tenant}/oauth2/authorize";
+
+        // The following cache could be private, but we keep it public so that internal unit test can take a peek into it
+        public static readonly ConcurrentDictionary<string, InstanceDiscoveryMetadataEntry> InstanceCache =
+            new ConcurrentDictionary<string, InstanceDiscoveryMetadataEntry>(); // Keys are host strings
+
+        private static SemaphoreSlim semaphore = new SemaphoreSlim(1, 1);
+
+        public static async Task<InstanceDiscoveryMetadataEntry> GetMetadataEntry(string host, bool validateAuthority,
+            CallState callState)
+        {
+            InstanceDiscoveryMetadataEntry entry = null;
+            if (!InstanceCache.TryGetValue(host, out entry))
+            {
+                await semaphore.WaitAsync().ConfigureAwait(false); // SemaphoreSlim.WaitAsync() will not block current thread
+                try
+                {
+                    if (!InstanceCache.TryGetValue(host, out entry))
+                    {
+                        await DiscoverAsync(host, validateAuthority, callState).ConfigureAwait(false);
+                        InstanceCache.TryGetValue(host, out entry);
+                    }
+                }
+                finally
+                {
+                    semaphore.Release();
+                }
+            }
+
+            return entry;
+        }
+
+        private static async Task DiscoverAsync(string host, bool validateAuthority,
+            CallState callState) // No return value. Modifies InstanceCache directly.
+        {
+            string tentativeAuthorizeEndpoint = AuthorizeEndpointTemplate.Replace("{host}", host);
+            string instanceDiscoveryEndpoint =
+                ("https://{host}/common/discovery/instance?api-version=1.1&authorization_endpoint=" +
+                 tentativeAuthorizeEndpoint)
+                .Replace("{host}", WhitelistedAuthorities.Contains(host) ? host : DefaultTrustedAuthority);
+            var client = new AdalHttpClient(instanceDiscoveryEndpoint, callState);
+            InstanceDiscoveryResponse discoveryResponse = null;
+            try
+            {
+                discoveryResponse = await client.GetResponseAsync<InstanceDiscoveryResponse>().ConfigureAwait(false);
+                if (validateAuthority && discoveryResponse.TenantDiscoveryEndpoint == null)
+                {
+                    // hard stop here
+                    throw new AdalException(AdalError.AuthorityNotInValidList);
+                }
+            }
+            catch (AdalServiceException ex)
+            {
+                if (validateAuthority)
```

Comment:
If validateAuthority is false and a 'Timeout' occurred, no exception will be thrown and ultimately the token acquisition will fail. Is there any value in throwing here for timeout?

Reply:
@kpanwar please confirm, I believe the validateAuthority process in ADAL is designed to be optional. If the developer CHOOSES to not require validation, the ADAL will just work using the authority specified by the developer. Subsequent token acquisition may or may not fail.

Reply:
@rayluo - you are correct. but if the validateAuthority is passed false, then how do you discover all the metadata? It seems like we will always validate authority for AAD from now on.

Reply:
@kpanwar The metadata is not a hard requirement. We will populate suitable default metadata and the normal workflow will continue. There is even a bunch of [metadata-less test cases](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/796/files#diff-34060c129eae8f8ac6dcaf1484ef10eaR47) to ensure that. Combining with that fact that you confirmed validateAuthority itself is also optional, I think we can conclude that this part of implementation is good.

Reply:
Any unexpected error is fatal error here. Otherwise this would result in inconsistent behavior on the client if for some reason the validation request failed due to timeout, but you continue in the flow and subsequent requests go through. You'll end up sending the request to the wrong host and doing token lookup and caching in the wrong place. This would result in inconsistent and bad behavior for users. It was much better to pass on the error that caused the authority validation to fail then to try to guess what caused the odd behavior later.